tools/kvmexit.py

```diff
@@ -0,0 +1,293 @@
+            if (tmp_info == NULL) {
+                return -1;
+            }
+        } else {
+            return -1;
+        }
+    }
+
+    if (er < REASON_NUM) {
+        tmp_info->exit_ct[er]++;
+        if (cache_miss == 1) {
+            if (cache_p->cache_pid_tgid != 0) {
+                // b.*.a Let's save the last hit cache_info into kvm_stat.
+                pcpu_kvm_stat.update(&cache_p->cache_pid_tgid, &cache_p->cache_exit_ct);
+            }
+            // b.* save to pcpu_cache
+            cache_p->cache_pid_tgid = cur_pid_tgid;
+            bpf_probe_read(&cache_p->cache_exit_ct, sizeof(*tmp_info), tmp_info);
+        }
+        return 0;
+    }
+
+    return -1;
+}
+"""
+
+# format output
+exit_reasons = (
+    "EXIT_REASON_EXCEPTION_NMI",
+    "EXIT_REASON_EXTERNAL_INTERRUPT",
+    "EXIT_REASON_TRIPLE_FAULT",
+    "EXIT_REASON_INIT_SIGNAL",
+    "",
+    "",
+    "",
+    "EXIT_REASON_PENDING_INTERRUPT",
+    "EXIT_REASON_NMI_WINDOW",
+    "EXIT_REASON_TASK_SWITCH",
+    "EXIT_REASON_CPUID",
+    "",
+    "EXIT_REASON_HLT",
+    "EXIT_REASON_INVD",
+    "EXIT_REASON_INVLPG",
+    "EXIT_REASON_RDPMC",
+    "EXIT_REASON_RDTSC",
+    "",
+    "EXIT_REASON_VMCALL",
+    "EXIT_REASON_VMCLEAR",
+    "EXIT_REASON_VMLAUNCH",
+    "EXIT_REASON_VMPTRLD",
+    "EXIT_REASON_VMPTRST",
+    "EXIT_REASON_VMREAD",
+    "EXIT_REASON_VMRESUME",
+    "EXIT_REASON_VMWRITE",
+    "EXIT_REASON_VMOFF",
+    "EXIT_REASON_VMON",
+    "EXIT_REASON_CR_ACCESS",
+    "EXIT_REASON_DR_ACCESS",
+    "EXIT_REASON_IO_INSTRUCTION",
+    "EXIT_REASON_MSR_READ",
+    "EXIT_REASON_MSR_WRITE",
+    "EXIT_REASON_INVALID_STATE",
+    "EXIT_REASON_MSR_LOAD_FAIL",
+    "",
+    "EXIT_REASON_MWAIT_INSTRUCTION",
+    "EXIT_REASON_MONITOR_TRAP_FLAG",
+    "",
+    "EXIT_REASON_MONITOR_INSTRUCTION",
+    "EXIT_REASON_PAUSE_INSTRUCTION",
+    "EXIT_REASON_MCE_DURING_VMENTRY",
+    "",
+    "EXIT_REASON_TPR_BELOW_THRESHOLD",
+    "EXIT_REASON_APIC_ACCESS",
+    "EXIT_REASON_EOI_INDUCED",
+    "EXIT_REASON_GDTR_IDTR",
+    "EXIT_REASON_LDTR_TR",
+    "EXIT_REASON_EPT_VIOLATION",
+    "EXIT_REASON_EPT_MISCONFIG",
+    "EXIT_REASON_INVEPT",
+    "EXIT_REASON_RDTSCP",
+    "EXIT_REASON_PREEMPTION_TIMER",
+    "EXIT_REASON_INVVPID",
+    "EXIT_REASON_WBINVD",
+    "EXIT_REASON_XSETBV",
+    "EXIT_REASON_APIC_WRITE",
+    "EXIT_REASON_RDRAND",
+    "EXIT_REASON_INVPCID",
+    "EXIT_REASON_VMFUNC",
+    "EXIT_REASON_ENCLS",
+    "EXIT_REASON_RDSEED",
+    "EXIT_REASON_PML_FULL",
+    "EXIT_REASON_XSAVES",
+    "EXIT_REASON_XRSTORS",
+    "",
+    "",
+    "EXIT_REASON_UMWAIT",
+    "EXIT_REASON_TPAUSE"
+)
+
+try:
+    has_kvm = subprocess.check_output("lsmod | grep kvm", shell=True)
+except Exception as e:
+    raise Exception("Please insmod kvm module to use kvmexit")
+
+try:
+    uname = subprocess.check_output("uname -r", shell=True).split('.')[:2]
+    if int(uname[0]) >= 5:
+        func_entry = "RAW_TRACEPOINT_PROBE(kvm_exit)"
+        get_er = "ctx->args[0]"
```

Comment:
Agree, please use `BPF.support_raw_tracepoint()`.

Reply:
OK, thanks for the suggestion. :)

Reply:
> BPF.support_raw_tracepoint()

How about the latter judgement on whether kernel version is newer than 4.7? Would it be more friendly to use `uname` to do the judgement and leave a message, or just omit the judgement like https://github.com/iovisor/bcc/blob/master/tools/runqlat.py#L207 does?

Reply:
> Use `BPF.support_raw_tracepoint()` instead.

How about the latter judgement on whether kernel version is newer than 4.7? Would it be more friendly to use uname to do the judgement and leave a message, or just omit the judgement like https://github.com/iovisor/bcc/blob/master/tools/runqlat.py#L207 does?

Reply:
> Use `BPF.support_raw_tracepoint()` instead.

Actually there are two stages to fully support raw_tracepoint: the first stage is only for in-kernel functions, and the second stage is for kernel modules and it is merged into kernel later than v5.0 (Ref: https://lore.kernel.org/patchwork/patch/1009374/). Considering kvm is a kernel module, maybe add a new function: **support_raw_tracepoint_module()**  to check this will be more appropriate? Just as follows: 
```
--- a/src/python/bcc/__init__.py
+++ b/src/python/bcc/__init__.py
@@ -1010,6 +1010,12 @@ class BPF(object):
             return True
         return False
    
    def support_raw_tracepoint_module():
        # kernel symbol "bpf_trace_modules" indicates raw_tracepoint module support
        if BPF.ksymname("bpf_trace_modules") != -1:
            return True
        return False
```

Reply:
> > Use `BPF.support_raw_tracepoint()` instead.
> 
> Actually there are two stages to fully support raw_tracepoint: the first stage is only for in-kernel functions, and the second stage is for kernel modules and it is merged into kernel later than v5.0 (Ref: https://lore.kernel.org/patchwork/patch/1009374/). Considering kvm is a kernel module, maybe add a new function: **support_raw_tracepoint_module()** to check this will be more appropriate? Just as follows:
> 
> ```
> --- a/src/python/bcc/__init__.py
> +++ b/src/python/bcc/__init__.py
> @@ -1010,6 +1010,12 @@ class BPF(object):
>              return True
>          return False
>     
>     def support_raw_tracepoint_module():
>         # kernel symbol "bpf_trace_modules" indicates raw_tracepoint module support
>         if BPF.ksymname("bpf_trace_modules") != -1:
>             return True
>         return False
> ```

Besides, how about the latter judgement on whether kernel version is older than 4.7? Would it be more friendly to use uname to do the judgement and leave a message, or just omit the judgement like https://github.com/iovisor/bcc/blob/master/tools/runqlat.py#L207 does?

Reply:
support_raw_tracepoint_module() sounds okay to me. We don't need to print out the message related to kernel version. Some kernels have backports, so in general, bcc tries very hard not to tie the feature with a particular kernel  version.